api/lib/infrastructure/repositories/campaign-assessment-participation-repository.js

```diff
@@ -0,0 +1,125 @@
+
+  const [campaignAssessmentParticipation] = await knex.with('campaignAssessmentParticipation',
+    (qb) => {
+      qb.select([
+        'users.id AS userId',
+        knex.raw('COALESCE ("schooling-registrations"."firstName", "users"."firstName") AS "firstName"'),
+        knex.raw('COALESCE ("schooling-registrations"."lastName", "users"."lastName") AS "lastName"'),
+        'campaign-participations.id AS campaignParticipationId',
+        'campaign-participations.campaignId',
+        'campaign-participations.createdAt',
+        'campaign-participations.sharedAt',
+        'campaign-participations.isShared',
+        'campaign-participations.participantExternalId',
+        'assessments.state',
+        _assessmentRankByCreationDate(),
+      ])
+        .from('campaign-participations')
+        .join('assessments', 'assessments.campaignParticipationId', 'campaign-participations.id')
+        .join('users', 'users.id', 'campaign-participations.userId')
+        .leftJoin('schooling-registrations', 'campaign-participations.userId', 'schooling-registrations.userId')
+        .leftJoin('campaigns', function() {
+          this.on({ 'campaign-participations.campaignId': 'campaigns.id' })
+            .andOn({ 'campaigns.organizationId': 'schooling-registrations.organizationId' });
+        })
+        .where({
+          campaignId,
+          'campaign-participations.id': campaignParticipationId
+        });
+    })
+    .from('campaignAssessmentParticipation')
+    .where({ rank: 1 });
+
+  if (campaignAssessmentParticipation == null) {
+    throw new NotFoundError(`There is no campaign participation with the id "${campaignParticipationId}"`);
+  }
+
+  return campaignAssessmentParticipation;
+}
+
+function _assessmentRankByCreationDate() {
+  return knex.raw('ROW_NUMBER() OVER (PARTITION BY ?? ORDER BY ?? DESC) AS rank', ['assessments.campaignParticipationId', 'assessments.createdAt']);
+}
+
+async function _buildCampaignAssessmentParticipation(result, targetedSkillIds) {
+  const testedSkillsCount = await _getTestedSkillsCountInTargetProfile(result, targetedSkillIds);
+  if (!result.isShared) {
+    return new CampaignAssessmentParticipation({
+      ...result,
+      totalSkillsCount: targetedSkillIds.length,
+      validatedSkillsCount: undefined,
+      testedSkillsCount,
+    });
+  }
+  return new CampaignAssessmentParticipation({
+    ...result,
+    totalSkillsCount: targetedSkillIds.length,
+    validatedSkillsCount: await _getValidatedSkillsCountInTargetProfile(result, targetedSkillIds),
+    testedSkillsCount,
+  });
+}
+
+async function _getValidatedSkillsCountInTargetProfile(result, targetedSkillIds) {
+  if (!result.isShared) {
+    return null;
+  }
+
+  const validatedTargetedSkillIdsByUser = await _getValidatedTargetSkillIds(
+    { [result.userId]: result.sharedAt },
+    targetedSkillIds,
+  );
+
+  return validatedTargetedSkillIdsByUser[result.userId] ? validatedTargetedSkillIdsByUser[result.userId].length : 0;
```

Comment:
Petite suggestion d'amélioration avec le get de lodash :
```suggestion
  return get(validatedTargetedSkillIdsByUser, `${result.userId}.length`, 0);
```
Cela peut être aussi appliqué à la ligne 100.